src/session/sidebar/mod.rs

```diff
@@ -201,4 +262,119 @@ impl Sidebar {
         self_.selected_chat.replace(selected_chat);
         self.notify("selected-chat");
     }
+
+    pub fn set_session(&self, session: Option<Session>) {
+        if self.session() == session {
+            return;
+        }
+
+        let self_ = imp::Sidebar::from_instance(self);
+
+        if let Some(ref session) = session {
+            // Merge ChatList and UserList into a single list model
+            let list = gio::ListStore::new(gio::ListModel::static_type());
+            list.append(session.chat_list());
+            list.append(session.user_list());
+            let model = gtk::FlattenListModel::new(Some(&list));
+
+            let filter = gtk::CustomFilter::new(
+                clone!(@strong self as obj => @default-return false, move |item| {
+                    let self_ = imp::Sidebar::from_instance(&obj);
+                    let is_searching = !self_.search_entry.text().is_empty();
+
+                    if is_searching {
+                        if let Some(chat) = item.downcast_ref::<Chat>() {
+                            self_.searched_chats.borrow().contains(&chat.id())
+                        } else if let Some(user) = item.downcast_ref::<User>() {
+                            // Show searched users, but only the ones that haven't
+                            // already been searched by the chats search
+                            !self_.already_searched_users.borrow().contains(&user.id())
+                                && self_.searched_users.borrow().contains(&user.id())
+                        } else {
+                            false
+                        }
+                    } else if let Some(chat) = item.downcast_ref::<Chat>() {
+                        chat.order() > 0
+                    } else {
+                        false
+                    }
+                }),
+            );
+            let sorter = gtk::CustomSorter::new(move |obj1, obj2| {
+                let chat1 = obj1.downcast_ref::<Chat>();
+                let chat2 = obj2.downcast_ref::<Chat>();
+
+                // Always show chats first and then users
+                if chat1.is_none() && chat2.is_some() {
```

Comment:
This would need fewer comparisions:
```rust
                if let Some(chat1) = chat1 {
                    if let Some(chat2) = chat2 {
                        chat1.order().cmp(&chat2.order()).into()
                    } else {
                        gtk::Ordering::Smaller
                    }
                } else if chat2.is_some() {
                    gtk::Ordering::Larger
                } else {
                    gtk::Ordering::Equal
                }
```